SFR-714 add blacklist functionality

mwbenowitz opened the pull request:
This updates the manner in which we are fetching ebook links through the OCLC lookup service, limiting it to a defined list of providers, chosen by the team.

This ensures that the project is not serving local links from libraries that are only available to their patrons, among other cases.

It does so by applying a regex for each approved provider and ensuring that the URI matches what we know to be an accessible link.

There is also a minor change to the fetching of publication dates and languages as that code existed in an un-committed state, and should be included here.

mwbenowitz committed:
SFR-714 Add improved 008 control field parsing

This is a dangling change from a fromer feature branch that adds a better way of pulling information from the `008` control field. Specifically this is the language and publication date data.

mwbenowitz committed:
SFR-714 Update ebook fetching from 856 field

Previously this field attempted to find and include any reference to an ebook link that was present in the `856` field. It did so by both applying a regex to the `u` subfield with the URI,  and the `k` and `q` subfields, which contain descriptions of the links. This instead applies the concept of a whitelist to only check for the presence of links from specific providers, namely: HathiTrust, InternetArchive and Project Gutenberg.

This checks a dict of regex values to evaluate the URI and if a match is found, store it. This has the side effect of simplifying the execution of this function and ensuring that external editions are only fetched in the expected situations

hokei commented on lib/parsers/parseOCLC.py:
```diff
@@ -218,41 +229,12 @@ def _matchRegexEbook(instance, uri, ebook_regex, uriIdentifier):
-            return True
-
-
-def _addEbook(instance, holding, subfield, uri, uriIdentifier):
```
Why is this not needed anymore?

mwbenowitz replied:
This was a method that was attempting to grab ebooks from anything that identified itself as an ebook. So it was looking in the URI description field for anything that contained "epub" or "ebook" and then added it as an Item regardless of the source. Since we only want these files from specific providers it is much easier to check that the URI matches what we expect since they have consistent formats.

This also has the side effect of excluding links from the providers that we want that don't match the expected format since we can't be sure that we would be able to display content from non-standard links

hokei commented on lib/parsers/parseOCLC.py:
```diff
@@ -186,7 +197,7 @@ def extractHoldingsLinks(holdings, instance):
 
 def _loadURIIdentifier(uri):
-    uriGroup = re.search(r'\/((?:(?!\/).)+)$', uri)
+    uriGroup = re.search(r'\/((?:(?!\/)[^.])+(?=$|\.[a-z]{3,4}$))', uri)
```
nit: I'd put a comment here on what this is doing for those who don't read regex well

mwbenowitz replied:
Good idea, I just had to look at it for a few minutes to remember exactly what it was doing

hokei commented on lib/parsers/parseOCLC.py:
```diff
@@ -186,7 +197,7 @@ def extractHoldingsLinks(holdings, instance):
 def _loadURIIdentifier(uri):
-    uriGroup = re.search(r'\/((?:(?!\/).)+)$', uri)
+    uriGroup = re.search(r'\/((?:(?!\/)[^.])+(?=$|\.[a-z]{3,4}$))', uri)
     if uriGroup is not None:
```
Q for my understanding: Is this the blacklist block?  So nothing happens if it matches the blacklist, and the format object with source/content_type/links gets added if it doesn't?

mwbenowitz replied:
The blacklist block is actually up at lines 159-162 where it sets a regex for each the providers we want to get ebook files from. This function helps pull useful identifiers from the URIs that we get for the files and passes the result back to the main function. The blacklist is then evaluated in `_matchRegexEbook` and if a match is found an `Item` is added to the instance being parsed.

hokei commented on tests/test_parseOCLC.py:
```diff
@@ -47,36 +42,6 @@ def test_holdings_bad_uri(self):
         extractHoldingsLinks([mock_holding], mock_instance)
         mock_add.assert_not_called()
 
```
Does there need to be another test for adding a book with a blacklisted uri?

mwbenowitz replied:
~A good idea~ actually this is what `test_match_ebook_skip` is doing, it attempts to add a URI that does not match the supplied regex and confirms that the `addFormat` method was not called.

mwbenowitz committed:
SFR-714 Add regex explanation

Adds a comment to clarify the id match regex and potentially help with future development to be sure we know what this method is doing
